MapboxNavigation/VanishingRouteLine.swift

```diff
@@ -0,0 +1,374 @@
+        guard let completeRoutePoints = routePoints else {
+            routeRemainingDistancesIndex = nil
+            return
+        }
+        let currentLegProgress = routeProgress.currentLegProgress
+        let currentStepProgress = routeProgress.currentLegProgress.currentStepProgress
+        /**
+        Find the count of remaining points in the current step.
+        */
+        var allRemainingPoints = getSlicedLinePointsCount(currentLegProgress: currentLegProgress, currentStepProgress: currentStepProgress)
+        
+        /**
+         Add to the count of remaining points all of the remaining points on the current leg, after the current step.
+        */
+        let currentLegSteps = completeRoutePoints.nestedList[routeProgress.legIndex]
+        let startIndex = currentLegProgress.stepIndex + 1
+        let endIndex = currentLegSteps.count - 1
+        if startIndex < endIndex {
+            allRemainingPoints += currentLegSteps.prefix(endIndex).suffix(from: startIndex).flatMap{ $0.compactMap{ $0 } }.count
+        }
+        
+        /**
+         Add to the count of remaining points all of the remaining legs.
+        */
+        for index in stride(from: routeProgress.legIndex + 1, to: completeRoutePoints.nestedList.count, by: 1) {
+            allRemainingPoints += completeRoutePoints.nestedList[index].flatMap{ $0 }.count
+        }
+        
+        /**
+         After calculating the number of remaining points and the number of all points,  calculate the index of the upcoming point.
+        */
+        let allPoints = completeRoutePoints.flatList.count
+        routeRemainingDistancesIndex = allPoints - allRemainingPoints - 1
+    }
+    
+    func getSlicedLinePointsCount(currentLegProgress: RouteLegProgress, currentStepProgress: RouteStepProgress) -> Int {
+        let startDistance = currentStepProgress.distanceTraveled
+        let stopDistance = currentStepProgress.step.distance
+        var slicedLine: Turf.LineString?
+        
+        /**
+         Implement the Turf.lineSliceAlong(lineString, startDistance, stopDistance) to return a sliced lineString.
+         */
+        if let lineString = currentStepProgress.step.shape {
+            if let midPoint = lineString.coordinateFromStart(distance: startDistance) {
+                slicedLine = lineString.trimmed(from: midPoint, distance: stopDistance - startDistance)
+            }
+        }
+        
+        if slicedLine != nil {
+            return slicedLine!.coordinates.count - 1
+        } else {
+            return 0
+        }
+    }
+    
+    func calculateRouteGranularDistances(coordinates: [CLLocationCoordinate2D]) -> RouteLineGranularDistances? {
+        if !coordinates.isEmpty {
+            return calculateGranularDistances(coordinates)
+        }
+        return nil
+    }
+    
+    func calculateGranularDistances(_ coordinates: [CLLocationCoordinate2D]) -> RouteLineGranularDistances {
```

Comment:
Shall we merge `calculateRouteGranularDistances` and `calculateGranularDistances` into one method? Verification which is performed in `calculateRouteGranularDistances` can simply move to `calculateGranularDistances`.

Reply:
Yes. Thanks for the suggestion. I'll work on it.